pkg/agent/credendtialvalidator.go

```diff
@@ -0,0 +1,153 @@
+package agent
+
+import (
+	"sync"
+	"time"
+
+	v1 "github.com/Axway/agent-sdk/pkg/apic/apiserver/models/api/v1"
+	management "github.com/Axway/agent-sdk/pkg/apic/apiserver/models/management/v1alpha1"
+	"github.com/Axway/agent-sdk/pkg/apic/provisioning"
+	"github.com/Axway/agent-sdk/pkg/config"
+	"github.com/Axway/agent-sdk/pkg/jobs"
+	"github.com/Axway/agent-sdk/pkg/util/log"
+)
+
+const (
+	status = "status"
+	state  = "state"
+)
+
+type cacheManager interface {
+	GetWatchResourceCacheKeys(group, kind string) []string
+	GetWatchResourceByKey(key string) *v1.ResourceInstance
+}
+
+type apicClient interface {
+	UpdateResourceInstance(ri v1.Interface) (*v1.ResourceInstance, error)
+	CreateSubResource(rm v1.ResourceMeta, subs map[string]interface{}) error
+}
+
+type credentialValidator struct {
+	jobs.Job
+	id           string
+	logger       log.FieldLogger
+	cacheManager cacheManager
+	client       apicClient
+}
+
+func newCredentialChecker(cacheManager cacheManager, client apicClient) *credentialValidator {
+	return &credentialValidator{
+		logger:       log.NewFieldLogger().WithComponent("credentialValidator"),
+		cacheManager: cacheManager,
+		client:       client,
+	}
+}
+
+// Ready -
+func (j *credentialValidator) Ready() bool {
+	return true
+}
+
+// Status -
+func (j *credentialValidator) Status() error {
+	return nil
+}
+
+// Execute -
+func (j *credentialValidator) Execute() error {
+	j.logger.Debug("validating credentials for expiration")
+
+	// Get all of the credentials from the cache
+	credKeys := j.cacheManager.GetWatchResourceCacheKeys(management.CredentialGVK().Group, management.CredentialGVK().Kind)
+
+	// loop all the keys in the cache and check if any have expired
+	now := time.Now()
+	wg := &sync.WaitGroup{}
+	for _, k := range credKeys {
+		wg.Add(1)
+		func(credKey string) {
```

Comment:
Guessing intent was to run this as a go routine?

Reply:
yes, thanks